### DISCUSSION TITLE
When writing migrations, how to write a "down" migration for a table drop?

### DISCUSSION THREAD

**Discussion by danya02** [SeaQL/sea-orm#1388] [Q&A] [answered]
I'm managing my database schema using `sea_orm_migration`. I've run into a situation where I'd like to drop an existing table. The "up" migration for this is obvious: use `SchemaManager::drop_table` on the name, which I can import from the migration script that created the table.

However, I don't really know what the best way to write the "down" migration is. Conceptually, a "down" migration should undo the action of an "up" migration, so that'd be to create the table in the state that it was before.

The two ways I can see to do this are:

- either write a `SchemaManager::create_table` statement that recreates the table in the state it was before the migration (but the downside is that I'd need to inspect the database to reconstruct the table state, which duplicates code as the code that generated the table can be found in older migrations)
- or construct and run one or more older "up" migrations that created the table in the first place (but then I'd have to search the migrations for all those that touched this table).

What is the recommended way to go about this?

(Btw, when should I write "down" migrations in general? The [`MigrationTrait` definition](https://github.com/SeaQL/sea-orm/blob/master/sea-orm-migration/src/lib.rs) only requires an implementation of the "up" migration, with the "down" migration erroring out by default -- does that mean that you're allowed to skip writing the "down" migration, and if so when?)

**Comment by billy1624** [comment, marked as answer]
Hey @danya02, welcome to SeaQL! This is a interesting question. IMO a "up" migration that drop a table is considered to be irreversible. I know this is against the idea of migration as a whole. In this case, I'd leave the "down" migration empty. That means going forward in migrations is still possible but you cannot go backward and cross the migration that does irreversible changes. Does that make sense to you?

**Reply by danya02** [reply]
Thanks for your reply. What would you say makes a migration irreversible -- is it that the operation involved (dropping a table, for example) destroys data?

Relatedly, how do you do migrations where a new table is populated with the results of a query on an old table? (Assuming such migrations are even a thing -- I'm assuming they are, because you might sometimes want to merge or split tables, but I'm very new to the concept of migrations in general so I don't know much about this.)

**Reply by billy1624** [reply]
> What would you say makes a migration irreversible -- is it that the operation involved (dropping a table, for example) destroys data?

Exactly!

> About copying data from an old to a new table

You can do it in three steps.

1. Create the new table
2. Copy the data from the old to the new table with `INSERT ... SELECT ... FROM` SQL statement
3. Make sure everything is fine before deleting the old table
